contracts/distribution.js

```diff
@@ -0,0 +1,301 @@
+/* eslint-disable no-await-in-loop */
+/* eslint no-underscore-dangle: ["error", { "allow": ["_id"] }] */
+/* global actions, api */
+
+// limit recipients per distribution batch
+const MAX_RECIPIENTS = 40;
+
+actions.createSSC = async () => {
+  const tableExists = await api.db.tableExists('batches');
+  if (tableExists === false) {
+    await api.db.createTable('batches');
+    await api.db.createTable('params');
+
+    const params = {};
+    params.distCreationFee = '500';
+    params.distUpdateFee = '250';
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  if (api.sender !== api.owner) return;
+
+  const {
+    distCreationFee,
+    distUpdateFee,
+  } = payload;
+
+  const params = await api.db.findOne('params', {});
+
+  if (distCreationFee) {
+    if (!api.assert(typeof distCreationFee === 'string' && !api.BigNumber(distCreationFee).isNaN() && api.BigNumber(distCreationFee).gte(0), 'invalid distCreationFee')) return;
+    params.distCreationFee = distCreationFee;
+  }
+  if (distUpdateFee) {
+    if (!api.assert(typeof distUpdateFee === 'string' && !api.BigNumber(distUpdateFee).isNaN() && api.BigNumber(distUpdateFee).gte(0), 'invalid distUpdateFee')) return;
+    params.distUpdateFee = distUpdateFee;
+  }
+
+  await api.db.update('params', params);
+};
+
+async function processBatch(batch, symbol, isFlush = false) {
+  let balance;
+  let balanceInd;
+  let payout;
+
+  for (let i = 0; i < batch.tokenBalances.length; i += 1) {
+    const curBalance = batch.tokenBalances[i];
+    if (curBalance.symbol === symbol) {
+      balanceInd = i;
+      balance = curBalance.quantity;
+      break;
+    }
+  }
+
+  for (let i = 0; i < batch.tokenMinPayout.length; i += 1) {
+    const minPayout = batch.tokenMinPayout[i];
+    if (minPayout.symbol === symbol) {
+      payout = minPayout.quantity;
+      break;
+    }
+  }
+
+  if (balance !== undefined && payout !== undefined
+    && (api.BigNumber(balance).gt(api.BigNumber(payout)) || isFlush === true)) {
+    // pay out token balance to recipients by configured share percentage
+    for (let i = 0; i < batch.tokenRecipients.length; i += 1) {
+      const recipient = batch.tokenRecipients[i];
+      const recipientShare = api.BigNumber(balance).multipliedBy(recipient.pct / 100).toFixed(3);
+      if (await api.transferTokens(recipient.account, symbol, recipientShare, recipient.type)) {
+        // eslint-disable-next-line no-param-reassign, max-len
+        batch.tokenBalances[balanceInd].quantity = api.BigNumber(batch.tokenBalances[balanceInd].quantity).minus(recipientShare);
+        await api.db.update('batches', batch);
```

Comment:
It's not necessary to update the DB on every pass through this for loop.  It's more efficient to do one single update at the end when the loop is complete.

Reply:
Ah yes. And this is related to the previous comment about storing the balance object. (It will be by reference so it will modify the existing batch object). You would be able to keep updating that and then do the db update at the end of the loop.

Reply:
Resolved by storing balance reference object.